src/iconforge.rs

```diff
@@ -0,0 +1,928 @@
+                        image::Rgba([0, 0, 0, 0])
+                    });
+                image::imageops::overlay(
+                    &mut blank_img,
+                    image,
+                    if x1 < 0 { (x1).abs() as i64 } else { 0 }
+                        - if x1 > i_width as i32 {
+                            (x1 - i_width as i32) as i64
+                        } else {
+                            0
+                        },
+                    if y1 < 0 { (y1).abs() as i64 } else { 0 }
+                        - if x1 > i_width as i32 {
+                            (x1 - i_width as i32) as i64
+                        } else {
+                            0
+                        },
+                );
+                *image = blank_img;
+                x1 = std::cmp::max(0, x1);
+                x2 = std::cmp::min(i_width as i32, x2);
+                y1 = std::cmp::max(0, y1);
+                y2 = std::cmp::min(i_height as i32, y2);
+                width = x2 - x1;
+                height = y2 - y1;
+            }
+            *image =
+                image::imageops::crop_imm(image, x1 as u32, y1 as u32, width as u32, height as u32)
+                    .to_image();
+        }
+    }
+    Ok(())
+}
+
+struct Rgba {
+    r: f32,
+    g: f32,
+    b: f32,
+    a: f32,
+}
+
+impl Rgba {
+    fn into_array(self) -> [u8; 4] {
+        [
+            self.r.round() as u8,
+            self.g.round() as u8,
+            self.b.round() as u8,
+            self.a.round() as u8,
+        ]
+    }
+
+    fn from_array(rgba: &[u8]) -> Rgba {
+        Rgba {
+            r: rgba[0] as f32,
+            g: rgba[1] as f32,
+            b: rgba[2] as f32,
+            a: rgba[3] as f32,
+        }
+    }
+
+    fn map_each(
+        color: Rgba,
+        color2: Rgba,
+        rgb_fn: &dyn Fn(f32, f32) -> f32,
+        a_fn: &dyn Fn(f32, f32) -> f32,
+    ) -> Rgba {
+        Rgba {
+            r: rgb_fn(color.r, color2.r),
+            g: rgb_fn(color.g, color2.g),
+            b: rgb_fn(color.b, color2.b),
+            a: a_fn(color.a, color2.a),
+        }
+    }
+
+    fn map_each_a(
+        color: Rgba,
+        color2: Rgba,
+        rgb_fn: &dyn Fn(f32, f32, f32, f32) -> f32,
+        a_fn: &dyn Fn(f32, f32) -> f32,
+    ) -> Rgba {
+        Rgba {
+            r: rgb_fn(color.r, color2.r, color.a, color2.a),
+            g: rgb_fn(color.g, color2.g, color.a, color2.a),
+            b: rgb_fn(color.b, color2.b, color.a, color2.a),
+            a: a_fn(color.a, color2.a),
+        }
+    }
+}
+
+fn blend_u8(color: &[u8], color2: &[u8], blend_mode: u8) -> [u8; 4] {
+    blend(
+        Rgba::from_array(color),
+        Rgba::from_array(color2),
+        blend_mode,
+    )
+    .into_array()
+}
+
+/// Blends two colors according to blend_mode. The numbers correspond to BYOND blend modes.
+fn blend(color: Rgba, color2: Rgba, blend_mode: u8) -> Rgba {
```

Comment:
not sure on the design of this, to me this seems like it should be on the rgba struct